clang/lib/CIR/CodeGen/CIRGenDecl.cpp

```diff
@@ -322,18 +333,277 @@ void CIRGenFunction::buildVarDecl(const VarDecl &D) {
+  else if (const auto *OMD = dyn_cast<ObjCMethodDecl>(DC))
+    llvm_unreachable("ObjC decl context for static var is NYI");
+  else
+    llvm_unreachable("Unknown context for static var decl");
+
+  ContextName += "." + D.getNameAsString();
+  return ContextName;
+}
+
+// TODO(cir): LLVM uses a Constant base class. Maybe CIR could leverage an
+// interface for all constants?
+mlir::cir::GlobalOp
+CIRGenModule::getOrCreateStaticVarDecl(const VarDecl &D,
+                                       mlir::cir::GlobalLinkageKind Linkage) {
+  // In general, we don't always emit static var decls once before we reference
+  // them. It is possible to reference them before emitting the function that
+  // contains them, and it is possible to emit the containing function multiple
+  // times.
+  if (mlir::cir::GlobalOp ExistingGV = StaticLocalDeclMap[&D])
+    return ExistingGV;
+
+  QualType Ty = D.getType();
+  assert(Ty->isConstantSizeType() && "VLAs can't be static");
+
+  // Use the label if the variable is renamed with the asm-label extension.
+  std::string Name;
+  if (D.hasAttr<AsmLabelAttr>())
+    Name = std::string(getMangledName(&D));
+  else
+    Name = getStaticDeclName(*this, D);
+
+  mlir::Type LTy = getTypes().convertTypeForMem(Ty);
+  assert(!UnimplementedFeature::addressSpace());
+
+  // OpenCL variables in local address space and CUDA shared
+  // variables cannot have an initializer.
+  mlir::Attribute Init = nullptr;
+  if (Ty.getAddressSpace() == LangAS::opencl_local ||
+      D.hasAttr<CUDASharedAttr>() || D.hasAttr<LoaderUninitializedAttr>())
+    llvm_unreachable("OpenCL & CUDA are NYI");
+  else
+    Init = builder.getZeroInitAttr(getTypes().ConvertType(Ty));
+
+  mlir::cir::GlobalOp GV = builder.createVersionedGlobal(
+      getModule(), getLoc(D.getLocation()), Name, LTy, false, Linkage);
+  // TODO(cir): infer visibility from linkage in global op builder.
+  GV.setVisibility(getMLIRVisibilityFromCIRLinkage(Linkage));
+  GV.setInitialValueAttr(Init);
+  GV.setAlignment(getASTContext().getDeclAlign(&D).getAsAlign().value());
+
+  if (supportsCOMDAT() && GV.isWeakForLinker())
+    llvm_unreachable("COMDAT globals are NYI");
+
+  if (D.getTLSKind())
+    llvm_unreachable("TLS mode is NYI");
+
+  setGVProperties(GV, &D);
+
+  // Make sure the result is of the correct type.
+  assert(!UnimplementedFeature::addressSpace());
+
+  // Ensure that the static local gets initialized by making sure the parent
+  // function gets emitted eventually.
+  const Decl *DC = cast<Decl>(D.getDeclContext());
+
+  // We can't name blocks or captured statements directly, so try to emit their
+  // parents.
+  if (isa<BlockDecl>(DC) || isa<CapturedDecl>(DC)) {
+    DC = DC->getNonClosureContext();
+    // FIXME: Ensure that global blocks get emitted.
+    if (!DC)
+      llvm_unreachable("address space is NYI");
+  }
+
+  GlobalDecl GD;
+  if (const auto *CD = dyn_cast<CXXConstructorDecl>(DC))
+    GD = GlobalDecl(CD, Ctor_Base);
```

Comment:
Are all these paths (ctor, dtor) exercised in tests? These should be asserts if not tested in this PR.